doc/source/guide/derivatives_guide.rst

```diff
@@ -0,0 +1,151 @@
+<https://en.wikipedia.org/wiki/Riesz_representation_theorem>`_, namely that the linear operator :math:`f'(x)` can be represented by a vector :math:`[\nabla f](x) \in \mathcal{X}` such that
+
+.. math::
+    f'(x)(y) = \langle y, [\nabla f](x) \rangle
+
+We call the (possibly nonlinear) operator :math:`x \rightarrow [\nabla f](x)` the *Gradient operator* of :math:`f`.
+
+Implementations in ODL
+~~~~~~~~~~~~~~~~~~~~~~
+
+* The gradient is implemented in ODL `Functional`'s via the `Functional.gradient` method.
+* It can be numerically computed using the `NumericalGradient` operator.
+
+Hessian
+#######
+In the classical setting of functionals :math:`f : \mathbb{R}^n \rightarrow \mathbb{R}`, the hessian in a point :math:`x` is the matrix :math:`H` such that
+
+.. math::
+    H(x) =
+    \begin{bmatrix}
+    \dfrac{\partial^2 f}{\partial x_1^2} & \dfrac{\partial^2 f}{\partial x_1\,\partial x_2} & \cdots & \dfrac{\partial^2 f}{\partial x_1\,\partial x_n} \\
+    \dfrac{\partial^2 f}{\partial x_2\,\partial x_1} & \dfrac{\partial^2 f}{\partial x_2^2} & \cdots & \dfrac{\partial^2 f}{\partial x_2\,\partial x_n} \\
+    \vdots & \vdots & \ddots & \vdots \\
+    \dfrac{\partial^2 f}{\partial x_n\,\partial x_1} & \dfrac{\partial^2 f}{\partial x_n\,\partial x_2} & \cdots & \dfrac{\partial^2 f}{\partial x_n^2}
+    \end{bmatrix}
+
+Where the derivatives are evaluated in the point :math:`x`. It has the property that that the quadratic variation of :math:`f` is
+
+.. math::
+    f(x + \Delta x) \approx f(x) + \langle \Delta x, [\nabla f](x)\rangle + \langle \Delta x, [H(x)](\Delta x)\rangle
+
+but also that the derivative of the gradient operator is
+
+.. math::
+    \nabla f(x + \Delta x) \approx [\nabla f](x) + [H(x)](\Delta x)
+
+If we take this second property as the *definition* of the Hessian, it can easily be generalized to the setting of functionals :math:`f : \mathcal{X} \rightarrow \mathbb{R}` mapping elements in some Hilbert space :math:`\mathcal{X}` to the real numbers.
+
+Implementations in ODL
+~~~~~~~~~~~~~~~~~~~~~~
+
+The hessian is not explicitly implemented anywhere in ODL, instead it is taken as the derivative of the gradient operator. This is however not implemented for all functionals.
+
+* For an example of a functional whose gradient has a derivative, see `RosenbrockFunctional`.
+* It can be computed by taking the `NumericalDerivative` of the gradient, which can in turn be computed using the `NumericalGradient`.
+
+Spatial Gradient
+################
+
+Thus the spatial gradient of the function :math:`f`, which is an element in some function space :math:`f \in \mathcal{F}(\Omega, \mathbb{R})`, is a element in the function space :math:`\mathcal{F}(\Omega, \mathbb{R}^n)` such that for any :math:`x, d \in \Omega`.
+
+.. math::
+    \lim_{h \rightarrow 0} \frac{\| f(x + h d) - f(x) - \langle h d, grad f \rangle \|}{h} = 0
```

Comment:
Leaving this one as is since this is probably more intuitive for most people and we really dont need frechet derivatives here.